src/compiler/checker.ts

```diff
@@ -6646,7 +6647,7 @@ namespace ts {
             }
 
             function hasExcessProperties(source: FreshObjectLiteralType, target: Type, reportErrors: boolean): boolean {
-                if (!(target.flags & TypeFlags.ObjectLiteralPatternWithComputedProperties) && maybeTypeOfKind(target, TypeFlags.ObjectType)) {
+                if (maybeTypeOfKind(target, TypeFlags.ObjectType) && !(target as ObjectType).inObjectLiteralPatternWithComputedProperties) {

```

Comment:
I was relying on unions and intersections not having `inObjectLiteralPatternWithComputedProperties` and having the check come back with undefined. Adding the check is safer, though.